src/lib/APS2Ethernet.cpp

```diff
@@ -0,0 +1,560 @@
+    typedef asio::ip::address_v4 addrv4;
+    asio::error_code ec;
+    addrv4 ip_addr = addrv4::from_string(serial, ec);
+    if (ec) {
+      FILE_LOG(logERROR) << "Invalid IP address: " << ec.message();
+      throw APS2_INVALID_IP_ADDR;
+    }
+    //Try the old port
+    //Put together the status request for old firmware devices
+    APS2EthernetPacket broadcastPacket = APS2EthernetPacket::create_broadcast_packet();
+    udp::endpoint endpoint(ip_addr, UDP_PORT_OLD);
+    udp_socket_old_.send_to(asio::buffer(broadcastPacket.serialize()), endpoint);
+    //And the new
+    endpoint.port(UDP_PORT);
+    uint8_t enumerate_request = 0x01;
+    udp_socket_.send_to(asio::buffer(&enumerate_request, 1), endpoint);
+
+    //Wait 100ms for response
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+    //Check again
+    if (devInfo_.find(serial) == devInfo_.end()) {
+      FILE_LOG(logERROR) << "APS2 failed to respond at " << serial;
+      throw APS2_NO_DEVICE_FOUND;
+    }
+  }
+
+  if (devInfo_[serial].supports_tcp) {
+    // C++14
+    // tcp_sockets_.insert(serial, std::make_unique<tcp::socket>(ios_));
+    // lowly C++11
+    std::unique_ptr<tcp::socket> sock(new tcp::socket(ios_));
+    asio::error_code ec;
+    FILE_LOG(logDEBUG1) << "Trying to connect to TCP port";
+    sock->connect(tcp::endpoint(asio::ip::address_v4::from_string(serial), TCP_PORT), ec);
+    if (ec) {
+      FILE_LOG(logERROR) << "Failed to connect to " << serial << " with error: " << ec.message();
+      throw APS2_SOCKET_FAILURE;
+    }
+
+    tcp_sockets_.insert(std::make_pair(serial, std::move(sock)));
+  } else {
+    msgQueue_lock_.lock();
+    msgQueues_[serial] = queue<APS2EthernetPacket>();
+    msgQueue_lock_.unlock();
+  }
+}
+
+void APS2Ethernet::disconnect(string serial) {
+  if (devInfo_[serial].supports_tcp) {
+    if (tcp_sockets_.find(serial) != tcp_sockets_.end()) {
+        tcp_sockets_[serial]->cancel();
+        tcp_sockets_[serial]->close();
+        tcp_sockets_.erase(serial);
+    }
+  } else {
+    msgQueue_lock_.lock();
+    msgQueues_.erase(serial);
+    msgQueue_lock_.unlock();
+  }
+}
+
+void APS2Ethernet::send(string ipAddr, const vector<APS2Datagram> & datagrams) {
+  FILE_LOG(logDEBUG2) << "APS2Ethernet::send";
+  if (devInfo_[ipAddr].supports_tcp) {
+    FILE_LOG(logDEBUG2) << "Sending " << datagrams.size() << " datagram" << (datagrams.size() > 1 ? "s" : "") << " over TCP";
+    // If we have TCP just send it out
+    for (const auto & dg : datagrams){
+      auto data = dg.data();
+      for (auto & val : data) {
+        val = htonl(val);
+      }
+      FILE_LOG(logDEBUG3) << "Sending datagram with command word " << hexn<8> << dg.cmd.packed <<
+        " to address " << hexn<8> << dg.addr << " with payload size " << std::dec << dg.payload.size() << " for total size " << data.size();
+      std::error_code ec;
+      asio::write(*tcp_sockets_[ipAddr], asio::buffer(data), ec); //TODO should this be async?
```

Comment:
`asio::write` is explicitly blocking and blocking till all the data is sent too.  But it actually would be a hassle to async send because you  have to keep the send buffer valid until the callback.